add Cody tutorial content to docs (#300)
sourcegraph/docs: docs/cody/use-cases/vsc-tutorial.mdx

# Cody Tutorial

> 

Reviewer: <p className="subtitle">This tutorial gives you a quick walkthrough of leveraging Cody for autocompletion, chat, explaining code, and running commands in VS Code.</p>

## Code Autocomplete

![](https://storage.googleapis.com/sourcegraph-assets/blog/vs-code-onboarding-walkthrough-dec-2023-cody-autocomplete-tsx.gif)

> Start writing code and Cody will complete the line (or the entire function) for you. Hit tab to accept the suggestion.

Reviewer: ```suggestion
Start writing code and Cody will complete the line (or the entire function) for you. Click **Tab** to accept the suggestion.
```

Author: Tab isn't a button - it's the "Tab" key on the keyboard. I'll change to "Press **Tab**"

**✨ Pro-tips for using Cody autocomplete**

• Autocomplete uses the surrounding code and context to inform the suggestions, so if you need to guide it you can add a comment above the line you're editing.

> 

Reviewer: ```suggestion
```
Rmv extra space between bullets

• You can hover over the grey suggestion to see a toolbar of alternative suggestions, as well as other options such as accepting a single word at a time.

> 

Reviewer: ```suggestion
```

• You can use the "Trigger Autocomplete at Cursor" command to trigger a code suggestion at any time, using the default keyboard shortcut of `Option` `\` on macOS and `Alt` `\` on Windows & Linux.

## Chat

![](https://storage.googleapis.com/sourcegraph-assets/blog/blog-vscode-v018-release/blog-v018-context-controls-002.gif)

Answer questions about general programming topics, or specific to your codebase, with Cody chat. You can choose between LLMs, @-mentions files and symbols, and enable enhanced repository-wide code context.

You can start a chat at any time using the default keyboard shortcut of `Option` `/` on macOS and `Alt` `/` on Windows & Linux.

**✨ Pro-tips for using Cody chat**

• Enabling “Enhanced Context”, and tagging specific files and symbols, gives the selected LLM more context to help answer your questions more accurately.

> 

Reviewer: ```suggestion
```

• If you want to chat directly with the LLM without any additional context adding, disable "Enhanced Context".

> 

Reviewer: ```suggestion
```

• You can increase the quality of your chat responses by enabling "Search Context (Beta)" in your [Cody Settings](command:cody.status-bar.interacted).

## Cody Commands

![](https://storage.googleapis.com/sourcegraph-assets/blog/vs-code-onboarding-walkthrough-dec-2023-explain.gif)

Cody has a range of commands for explaining code, generating unit tests, adding documentation, and more.

> To get started: select code in your editor, right click, and choose a command from the "Cody" menu.

Reviewer: ```suggestion
To get started, select code in your editor, right-click, and choose a command from the "Cody" menu.
```

You can also use the [Cody: Commands Menu](command:cody.menu.commands) which has the default keyboard shortcut of `Option` `C` on macOS and `Alt` `C` on Windows & Linux.

**✨ Pro-tips for using Cody commands**

> • You can build your own [Custom Commands (Beta)](https://sourcegraph.com/docs/cody/custom-commands) with custom prompts, output into chat or perform code edits, and more.

Reviewer: ```suggestion
• You can build your own [Custom Commands (Beta)](/cody/capabilities/custom-commands) with custom prompts, output into chat or perform code edits, and more.
```
Add relative paths for links

## Edit Code

You can ask Cody to perform code edits using the Edit Code command. This will edit code according to your instructions, giving you the chance to retry the edit with a new prompt, undo, or accept the change.

> To get started: select code in your editor, right click, and choose "Cody → Edit Code".

Reviewer: ```suggestion
To get started, select code in your editor, right-click, and choose "Cody > Edit Code".
```

You can also use the default keyboard shortcut of `Option` `K` on macOS and `Alt` `K` on Windows & Linux.

**✨ Pro-tips for code editing with Cody**

• You can open the 💡 menu, with an "Cody: Edit Code" option, by selecting any line of code and using the keyboard short `Command` `.` on macOS or `Crtl` `.` on Windows & Linux.

> 

Reviewer: ```suggestion
```

• If you start an empty line and open the 💡 menu (using the keyboard short `Command` `.` on macOS or `Crtl` `.` on Windows & Linux) you can use "Cody: Generate Code" to generate new lines of code based on instructions.

> • You define your own custom code editing commands using [Custom Commands (Beta)](https://sourcegraph.com/docs/cody/custom-commands), by setting the [`commands.<id>.mode` property](https://sourcegraph.com/docs/cody/capabilities/commands#commands-id-mode).

Reviewer: ```suggestion
• You define your own custom code editing commands using [Custom Commands (Beta)](/cody/capabilities/custom-commands), by setting the [`commands.<id>.mode` property](/cody/capabilities/commands).
```

## Explain Code

> <video autoPlay muted loop playsInline>

Reviewer: ```suggestion
<video width="1920" height="1080" loop playsInline controls style={{ width: '100%', height: 'auto' }}>
 <source src="https://storage.googleapis.com/sourcegraph-assets/website/Product%20Animations/cody-explain-code-aug2023.mp4" type="video/mp4"/>
</video>
```

The correct format to add a video component.

>     <source

Reviewer: ```suggestion
```

>         type="video/mp4"

Reviewer: ```suggestion
```

>         src="https://storage.googleapis.com/sourcegraph-assets/website/Product%20Animations/cody-explain-code-aug2023.mp4"

Reviewer: ```suggestion
```

>     />

Reviewer: ```suggestion
```

> </video>

Reviewer: ```suggestion
```

Use Cody to get an in-depth explanation of any piece of code in any programming language.

> To get started: select code in your editor, right click, and choose "Cody → Explain Code".

Reviewer: ```suggestion
To get started, select code in your editor, right-click, and choose "Cody > Explain Code".
```

You can also run this command from the [Cody: Commands Menu](command:cody.menu.commands), which has the default keyboard shortcut of `Option` `C` on macOS and `Alt` `C` on Windows & Linux.

**✨ Pro-tips for understanding code with Cody**

• Cody can also explain errors too, with the "Ask Cody to Explain" option in the 💡 menus, or by right clicking on any items in the "Problems" tab of VS Code.

> 

Reviewer: ```suggestion
```

> • You can define your own custom code explain commands to suit, such as prompt that requests an explanation focused on potential security issues, using [Custom Commands (Beta)](https://sourcegraph.com/docs/cody/custom-commands).

Reviewer: ```suggestion
• You can define your own custom code to explain commands to suit, such as a prompt that requests an explanation focused on potential security issues, using [Custom Commands (Beta)](/cody/capabilities/custom-commands).
```

## Fix Code

![](https://storage.googleapis.com/sourcegraph-assets/blog/vs-code-onboarding-walkthrough-dec-2023-ask-to-fix.gif)

Something wrong with your code? Use Cody’s "Ask Cody to Fix" command to fix it, or explain the problem.

**✨ Pro-tips for fixing code with Cody**

• You can open the 💡 menu, with an "Ask Cody to Fix" option, by moving the cursor over any line of code with an error and using the keyboard short `Command` `.` on macOS or `Crtl` `.` on Windows & Linux.

> 

Reviewer: ```suggestion
```

2. In the `Keyboard Shortcuts Editor`, you can search for a specific command by typing in the search bar.
3. Once you have found the command you want to reassign keybindings for, click on it to select it.
4. Follow the instruction on screen to assign a new keybinding.

**✨ Pro-tips for assigning keyboard shortcuts for Cody**

• You can assign key bindings (keyboard shortcuts) to individual custom commands.

## Natural Language Search

![](https://storage.googleapis.com/sourcegraph-assets/blog/vs-code-onboarding-walkthrough-dec-2023-natural-language.gif)

Cody builds a Search index of your local files to make it easier to find what you’re looking for. Use a natural language query like “password hashing” or "connection retries" to quickly find and open the files that match your search.

**✨ Pro-tips for using Cody’s Natural Language Search**

• Increase the quality of your chat responses by enabling "Search Context (Beta)" in your [Cody Settings](command:cody.status-bar.interacted). This enables chat to use the natural language search index as a context source.

> 

Reviewer: ```suggestion
```
